fix #2338 Add new Emission FAIL_ZERO_SUBSCRIBER error code

simonbasle opened the pull request.

bsideup commented on reactor-core/src/main/java/reactor/core/publisher/EmitterProcessor.java:
```diff
@@ -310,7 +310,7 @@ public Emission tryEmitNext(T t) {
 
 		if (!q.offer(t)) {
-			return Emission.FAIL_OVERFLOW;
+			return subscribers == EMPTY ? Emission.FAIL_ZERO_SUBSCRIBER : Emission.FAIL_OVERFLOW;
```
shouldn't we check `subscribers == EMPTY` even before attempting to lazily create a queue? Otherwise, it still returns `OK` if the queue has the capacity but there are no subscribers

simonbasle replied:
One of the advertised features of the `EmitterProcessor` is the capacity to buffer a (few?) elements into that queue before failing with overflow, including when trying to emit before anybody's listening. If the queue has the capacity but there are no listeners, the first listener will get the early emission from the queue, so this is intended.

bsideup commented on reactor-core/src/main/java/reactor/core/publisher/Sinks.java:
```diff
@@ -497,6 +502,8 @@ public Emission getReason() {
 	 */
 	public interface Many<T> {
 
+		boolean hasSubscriber();
```
Have you considered naming it `hasSubscribers`?

simonbasle replied:
actually, since there's no direct need for it in this PR it should probably be debated in a separate issue 🤔

bsideup replied:
Sounds good! Less is better :)

bsideup commented on reactor-core/src/main/java/reactor/core/publisher/UnicastManySinkNoBackpressure.java:
```diff
@@ -67,6 +70,18 @@
 		Objects.requireNonNull(actual, "subscribe");
 
+		//detect pre-subscribe termination
+		if (STATE.compareAndSet(this, State.TERMINATED_BEFORE_SUBSCRIPTION, State.TERMINATED)) {
```
to avoid double CASing, we can/should replace it with `do { read + switch + cas } while (casFails)`

simonbasle replied:
done, let me know if that's what you had in mind

bsideup commented on reactor-core/src/main/java/reactor/core/publisher/UnicastManySinkNoBackpressure.java:
```diff
@@ -153,16 +169,24 @@ public void emitError(Throwable error) {
+					this.error = t;
+					return Emission.OK;
+				}
+				else return tryEmitError(t);
```
```suggestion
				else {
					return tryEmitError(t);
				}
```

bsideup replied:
Not resolved.

simonbasle replied:
done in eca5e18

bsideup commented on reactor-core/src/main/java/reactor/core/publisher/UnicastManySinkNoBackpressure.java:
```diff
@@ -182,18 +206,29 @@ public void emitComplete() {
+					actual = null;
+					return Emission.OK;
+				}
+				else return tryEmitComplete(); // recurse
```
```suggestion
				else {
					return tryEmitComplete(); // recurse
				}
```

bsideup commented on reactor-core/src/test/java/reactor/core/publisher/UnicastProcessorTest.java:
```diff
@@ -279,4 +279,50 @@ public void shouldSignalErrorOnOverflow() {
+	public void tryEmitNextWithNoSubscriberAndBoundedQueueFailsZeroSubscriber() {
+		UnicastProcessor<Integer> unicastProcessor = UnicastProcessor.create(Queues.<Integer>one().get());
+
+		assertThat(unicastProcessor.tryEmitNext(1)).isEqualTo(Sinks.Emission.OK);
```
IMO this is wrong and should return `FAIL_ZERO_SUBSCRIBER`

simonbasle replied:
the fact that `UnicastProcessor` allows to pre-emit (a few) elements before the actual subscription occurs is IMHO a cornerstone of it working so well as the plumbing for widowing operators. I fear doing what you suggest would open the door to all sorts of issues in windowing operators, at least.

bsideup replied:
if so, the second emission should fail with OVERFLOW and not "zero subscribers" because the sink/processor isn't subscriber-dependent

simonbasle replied:
why? the whole point of adding this failed state is to indicate _undeliverable_ emission (as opposed to deliverable-in-case-of-later-subscription). there are still 0 subscribers when the second `tryEmitNext` occurs, but that one cannot be buffered for the benefit of a deferred subscriptions, so we return `FAIL_ZERO_SUBSCRIBER` (to distinguish from "there's a subscriber but it hasn't requested enough").

```java
		/**
		 * Has failed to emit the signal because the sink has never been subscribed to [AND]
         * has no capacity to buffer the signal.
		 */
		FAIL_ZERO_SUBSCRIBER;
```

(I should change the above wording to say `has currently no subscriber` maybe rather than `has never been subscribed)

bsideup replied:
IMO the failure's cause is the overflow. The fact that it overflows when there are no subscribers is an implementation detail. Getting `FAIL_ZERO_SUBSCRIBER` after `OK` is super weird and IMO counterintuitive 😅

bsideup commented on reactor-core/src/main/java/reactor/core/publisher/UnicastManySinkNoBackpressure.java:
```diff
@@ -67,13 +70,38 @@
+		} while (casFails);
+
+		if (s == State.TERMINATED_BEFORE_SUBSCRIPTION) {
+			//detected pre-subscribe termination
+			if (this.error != null) {
+				Operators.error(actual, this.error);
+			}
+			else {
+				Operators.complete(actual);
+			}
```
why not put it into switch's cases?
Similar to https://github.com/reactor/reactor-core/blob/56e5c0ddf8e9485b9117c75795db0afcd64eef6e/reactor-core/src/main/java/reactor/core/publisher/FluxConcatMapNoPrefetch.java#L276-L286

1. try CAS
2. continue if fails
3. do the action otherwise

simonbasle replied:
see 80278ec

simonbasle commented:
@bsideup fixed the last two comments + rebased on top of master. Once approved I will probably rework the commits into a polish commit for `UnicastManySinkNoBackpressure` handling of pre-subscribe terminations and one commit for the actual `FAIL_ZERO_SUBSCRIBER`

bsideup commented on reactor-core/src/main/java/reactor/core/publisher/UnicastManySinkNoBackpressure.java:
```diff
@@ -67,13 +70,34 @@
-		if (!STATE.compareAndSet(this, State.INITIAL, State.SUBSCRIBED)) {
-			Operators.reportThrowInSubscribe(actual, new IllegalStateException("Unicast Sinks.Many allows only a single Subscriber"));
-			return;
+		State s;
```
nit: Could use a similar style as in https://github.com/reactor/reactor-core/blob/56e5c0ddf8e9485b9117c75795db0afcd64eef6e/reactor-core/src/main/java/reactor/core/publisher/FluxConcatMapNoPrefetch.java#L226

bsideup commented on reactor-core/src/main/java/reactor/core/publisher/UnicastManySinkNoBackpressure.java:
```diff
@@ -153,16 +178,26 @@ public void emitError(Throwable error) {
 			case INITIAL:
-				// TODO different Emission?
-				return Emission.OK;
+				if (STATE.compareAndSet(this, State.INITIAL, State.TERMINATED_BEFORE_SUBSCRIPTION)) {
```
consider extracting `this.state` into a variable and using it here in place of `State.INITIAL` to avoid future copy/paste errors

bsideup commented:
@simonbasle I am still feeling very uncomfortable with https://github.com/reactor/reactor-core/pull/2357/files#r484884433, but you can veto it :)

simonbasle commented:
extracted the polishing of `UnicastManySinkNoBackpressure` unrelated to 0-sub in #2367. I squashed the rest of the previous 14 commits into one + added a new commit for changes around the way `emitNext` deals with `FAIL_ZERO_SUBSCRIBER` (hint: it cannot really do much, but now at least it doesn't error and so doesn't terminate the sink).

bsideup reviewed (approved):
prematurely approving, but `FAIL_ZERO_SUBSCRIBER` should also be returned from `tryEmitError`/`tryEmitComplete`.
